### DISCUSSION TITLE
Would a print("Plain Text") function be useful?

### DISCUSSION THREAD

**Discussion by cajhin** [cajhin/capsicain#16] [Ideas]
Right now you cannot sent text, only key sequences. Those are very flexible, but a little awkward.

To output "Hello" you need (some combo) >  sequence ( &LSHF_H_^LSHF_E_L_L_O )
Would be easier to just write (some combo) > print ( Hello )

capsicain can only send key events, not strings.
Capital H needs to be translated to "Shift down, h, Shift up". No problem.

Trouble starts with chars like '&'. There is no & key. The combo to create it depends on your language settings.
In US layout, Shift+7 = &
In DE layout, Shift+6 = &

I could:
1. check the Windows language and have 1000 translation rules for all kind of languages (nope!)
2. translate all special chars to altNumpad combos (works but ugly and fails in Linux VMs)
3. Simply disallow anything that is not A-Z / space / enter / . / ,
4. Leave it as it is: you want to output a string, define it manually (just record a macro and paste that into the ini)

I don't really need it myself, but I'm open to ideas.

**Comment by bitoj** [comment]
I don't really need it, but it might be convenient at some point, especially when it helps me enter multilingual text. ASCII sequences are easy enough right now.

However, as far as I can tell, none of your proposed solutions will work in all cases, because there is no general solution (that I know of) for entering all supported unicode codepoints for all users in all applications using the keyboard.

- Somehow translating back characters to scancodes using the current keyboard lay-out limits you to the primary locale of the system.
- Using alt codes between 128 and 255 without a leading 0 limits you to the OEM codepage for the current locale.
- Using alt codes between 128 and 255 with a leading 0 limits you to the Window ("ANSI") codepage for the current locale (which might not the same as in Windows versions for North-America and Western Europe).
- Support for decimal alt codes above 255 for unicode code points is application dependent.
- Currently available ways to enter a hexadecimal representation for unicode code points require either a special registry entry or application support (or maybe both).

If my list is complete, there is no general algorithmic mapping from characters to key sequences. The solution will then depend on the user locale and other aspects of the Windows configuration. Which suggests using your option 4: a macro will probably be replayed in the same environment where it was recorded.

I hope someone can correct me if I missed something.
